File: cli/cmd/export.go
package cmd

import (
	"bytes"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	simkubev1 "simkube/lib/go/api/v1"
	"simkube/lib/go/util"
)

const (
	subcmdName = "export"

	startTimeFlag          = "start-time"
	endTimeFlag            = "end-time"
	excludedNamespacesFlag = "excluded-namespaces"
	excludedLabelsFlag     = "excluded-labels"
	outputFlag             = "output"
	tracerAddrFlag         = "tracer-addr"
)

func Export() *cobra.Command {
	export := &cobra.Command{
		Use:   subcmdName,
		Short: "export trace data",
		Run:   doExport,
	}
	export.Flags().String(
		startTimeFlag,
		"-30m",
		"start time; can be a relative duration or absolute (local) timestamp\n"+
			"    in ISO-8601 extended format (YYYY-MM-DDThh:mm:ss).\n"+
			"    durations are computed relative to the specified end time,\n"+
			"    _not_ the current time\n",
	)
	export.Flags().String(endTimeFlag, "now", "end time; can be a relative or absolute (local) timestamp\n")
	export.Flags().StringArray(
		excludedNamespacesFlag,
		[]string{"kube-system", "monitoring", "local-path-storage", "simkube", "cert-manager", "volcano-system"},
		"namespaces to exclude from the trace\n",
	)
	export.Flags().StringArray(
		excludedLabelsFlag,
		[]string{},
		"label selectors to exclude from the trace (key=value pairs)",
	)

	export.Flags().String(tracerAddrFlag, "http://localhost:7777", "tracer server address\n")
	export.Flags().StringP(outputFlag, "o", "file:///tmp/kind-node-data", "location to save exported trace\n")
	return export
}

func doExport(cmd *cobra.Command, _ []string) {
	// None of these error conditions should get hit, since they are all assigned default values?
	// I'm not sure if there's a better way to do this or not.
	startTimeStr, err := cmd.Flags().GetString(startTimeFlag)
	if err != nil {
		fmt.Printf("no start time flag: %v\n", err)
		os.Exit(1)
	}
	endTimeStr, err := cmd.Flags().GetString(endTimeFlag)
	if err != nil {
		fmt.Printf("no end time flag: %v\n", err)
		os.Exit(1)
	}
	excludedNamespaces, err := cmd.Flags().GetStringArray(excludedNamespacesFlag)
	if err != nil {
		fmt.Printf("no namespaces flag: %v\n", err)
		os.Exit(1)
	}
	tracerAddr, err := cmd.Flags().GetString(tracerAddrFlag)
	if err != nil {
		fmt.Printf("no tracer-addr flag: %v\n", err)
		os.Exit(1)
	}
	output, err := cmd.Flags().GetString(outputFlag)
	if err != nil {
		fmt.Printf("no output flag: %v\n", err)
		os.Exit(1)
	}

	// TODO actually parse excluded labels
	// excludedLabels, _ := cmd.Flags().GetStringArray(excludedLabelsFlag)

	endTime, err := util.ParseTimeStr(endTimeStr, time.Time{})
	if err != nil {
		fmt.Printf("could not parse end time: %v", err)
		os.Exit(1)
	}
	startTime, err := util.ParseTimeStr(startTimeStr, endTime)
	if err != nil {
		fmt.Printf("could not parse start time: %v", err)
		os.Exit(1)
	}

	filters := *simkubev1.NewExportFilters(
		excludedNamespaces,
		[]metav1.LabelSelector{},
		true,
	)
	request := simkubev1.NewExportRequest(startTime.Unix(), endTime.Unix(), filters)
	requestJSON, err := request.MarshalJSON()
	if err != nil {
		fmt.Printf("could not marshal request to JSON: %v\n", err)
		os.Exit(1)
	}

	requestBody := bytes.NewReader(requestJSON)

	exportUrl := fmt.Sprintf("%s/export", tracerAddr)
	fmt.Println("exporting trace data")
	fmt.Printf("start_ts = %v, end_ts = %v\n", startTime, endTime)
	fmt.Printf("using filters:\n\texcluded_namespaces: %v\n\texcluded_labels: none\n", excludedNamespaces)
	fmt.Printf("making request to %s\n", exportUrl)

	req, err := http.NewRequest(http.MethodPost, exportUrl, requestBody)
	if err != nil {
		fmt.Printf("could not create request: %v\n", err)
		os.Exit(1)
	}

	//nolint:bodyclose // this gets closed at the end of the function anyways it's fine NBD
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Printf("error making request: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("got response status: %d\n", resp.StatusCode)
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("could not read response body: %v\n", err)
		os.Exit(1)
	}

	if err = writeOutput(output, respBody); err != nil {
		fmt.Printf("could not write trace data to %s: %v\n", output, err)
		os.Exit(1)
	}
}

func writeOutput(output string, data []byte) error {
	if !strings.HasPrefix(output, "file://") {
		return fmt.Errorf("only local output locations supported: %s", output)
	}

	location := strings.TrimPrefix(output, "file://")
	if err := os.MkdirAll(location, fs.ModeDir|0755); err != nil {
		return fmt.Errorf("could not create location %s: %w", location, err)
	}
	fullname := fmt.Sprintf("%s/trace", location)
	out, err := os.Create(fullname)
	if err != nil {
		return fmt.Errorf("could not open %s for writing: %w", fullname, err)
	}
	defer func() {
		if err := out.Close(); err != nil {
			panic(err)
		}
	}()

	if _, err = out.Write(data); err != nil {
		return fmt.Errorf("could not write data to %s: %w", location, err)
	}
	fmt.Printf("trace successfully stored to %s\n", output)
	return nil
}

File: cli/export.rs
use anyhow::bail;
use simkube::prelude::*;
use simkube::store::external_storage::{
    ObjectStoreWrapper,
    SkObjectStore,
};
use simkube::time::duration_to_ts;

#[derive(clap::Args)]
pub struct Args {
    #[arg(
        short,
        long,
        long_help = "trace export start timestamp; can be a relative duration\n\
             or absolute timestamp; durations are computed relative\n\
             to the specified end time, _not_ the current time",
        default_value = "-30m",
        value_parser = duration_to_ts,
        allow_hyphen_values = true,
    )]
    pub start_time: i64,

    #[arg(
        short = 't',
        long,
        long_help = "end time; can be a relative or absolute timestamp",
        default_value = "now",
        value_parser = duration_to_ts,
        allow_hyphen_values = true,
    )]
    pub end_time: i64,

    #[arg(
        long,
        long_help = "namespaces to exclude from the trace",
        value_delimiter = ',',
        default_value = "cert-manager,kube-system,local-path-storage,monitoring,simkube"
    )]
    pub excluded_namespaces: Vec<String>,

    #[arg(
        long,
        long_help = "sk-tracer server address",
        default_value = "http://localhost:7777"
    )]
    pub tracer_address: String,

    #[arg(
        short,
        long,
        long_help = "location to save exported trace",
        default_value = "file:///tmp/kind-node-data"
    )]
    pub output_path: String,
}

pub async fn cmd(args: &Args) -> EmptyResult {
    let filters = ExportFilters::new(args.excluded_namespaces.clone(), vec![], true);
    let req = ExportRequest::new(args.start_time, args.end_time, args.output_path.clone(), filters);
    let endpoint = format!("{}/export", args.tracer_address);

    println!("exporting trace data");
    println!("start_ts = {}, end_ts = {}", args.start_time, args.end_time);
    println!("using filters:\n\texcluded_namespaces: {:?}\n\texcluded_labels: none", args.excluded_namespaces);
    println!("making request to {}", endpoint);

    let client = reqwest::Client::new();
    match client.post(endpoint).json(&req).send().await? {
        res if res.status().is_success() => {
            // If we got trace data back from the request, it means the tracer pod couldn't or
            // didn't want to write it (e.g., we asked to write to a local file); in the future we
            // might also try to write the data to the cloud provider storage as a fallback if it
            // didn't work from the tracer pod, so this will handle that case as well.
            let data = res.bytes().await?;
            if !data.is_empty() {
                let object_store = SkObjectStore::new(&args.output_path)?;
                object_store.put(data).await?;
            }
            println!("Trace data exported to {}", args.output_path);
        },
        res => bail!("Received {} response; could not export trace data:\n\n{}", res.status(), res.text().await?),
    };
    Ok(())
}
